### PR TITLE
Adding diagnostics tools for ill conditioning

### PR DESCRIPTION
**Pull request by andrewlee94**
## Summary/Motivation:
This PR adds new diagnostics capabilities for detecting near-parallel variables and constraints and for finding sources of ill conditioning in models (similar to DegeneracyHunter).

## Changes proposed in this PR:
- Adds method to check for near parallel rows and columns in Jacobians
- Add hook for checking for parallel rows and columns in DiagnosticsToolbox
- Adds beta method to analyze Jacobian for near singularity based on rows or columns (with logger warning about future changes)

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: idaes/core/util/model_diagnostics.py
Comment (bknueven) [inline idaes/core/util/model_diagnostics.py:18]:
```suggestion
__author__ = "Alexander Dowling, Douglas Allan, Andrew Lee, Robby Parker, Ben Knueven"
```

File: idaes/core/util/model_diagnostics.py
Comment (bknueven) [inline idaes/core/util/model_diagnostics.py:3303]: After using this on some larger flowsheets, I found the model build time to be fairly long. We could probably significantly speed it up by iterating over the non-zeros in the Jacobian instead.

We'd need to re-work the logic around selecting the row or column to do so.

File: idaes/core/util/model_diagnostics.py
Comment (bknueven) [inline idaes/core/util/model_diagnostics.py:3331]:
```suggestion
    results = solver.solve(inf_prob, tee=False)
    if check_optimal_termination(results):
        # TODO: maybe we should tighten tolerances first?
        raise RuntimeError("Ill conditioning diagnostic problem infeasible") 
```
The problem by construction should not be infeasible, but I am sure it the solver could report it as such for a sufficiently ill-conditioned system.

File: idaes/core/util/model_diagnostics.py
Comment (bknueven) [inline idaes/core/util/model_diagnostics.py:3333]: I have seen cases where this number is slightly negative. In this case we should probably bail out, or tighten the tolerances and try again (and if that fails then bail out).
```suggestion
    result_norm = inf_prob.res_norm.value
    if result_norm < 0.0:
        # TODO: try again with tighter tolerances?
        raise RuntimeError("Ill conditioning diagnostic problem has numerically troublesome solution")
```

File: idaes/core/util/model_diagnostics.py
Comment (bknueven) [inline idaes/core/util/model_diagnostics.py:3343]:
```suggestion
    # if the this problem is numerically infeasible, we can still report something to the user
    results = solver.solve(inf_prob, tee=True, load_solutions=False)
    if check_optimal_termination(results):
        inf_prob.solutions.load_from(results)
```

File: idaes/core/util/model_diagnostics.py
Comment (bknueven) [inline idaes/core/util/model_diagnostics.py:3263]:
```suggestion
    # Get a mapping of row/column names to index in jac
    con_dict = {i.name: idx for idx, i in enumerate(nlp.get_pyomo_constraints())}
    var_dict = {i.name: idx for idx, i in enumerate(nlp.get_pyomo_variables())}
```

File: idaes/core/util/model_diagnostics.py
Comment (bknueven) [inline idaes/core/util/model_diagnostics.py:3291]:
```suggestion
                iidx = var_dict[i]
                jidx = con_dict[j]
```

File: idaes/core/util/model_diagnostics.py
Comment (bknueven) [inline idaes/core/util/model_diagnostics.py:3295]:
```suggestion
                jidx = var_dict[j]
                iidx = con_dict[i]
```

File: idaes/core/util/model_diagnostics.py
Comment (bknueven) [inline idaes/core/util/model_diagnostics.py:3268]:
```suggestion
    inf_prob.con_set = Set(initialize=con_dict.keys())
    inf_prob.var_set = Set(initialize=var_dict.keys())
```

**Review by bknueven — COMMENTED** [review]
The below set of changes avoids cubic time complexity in the `check_ill_conditioning` model construction.

File: idaes/core/util/tests/test_model_diagnostics.py
Comment (Robbybp) [inline idaes/core/util/tests/test_model_diagnostics.py:3106]: Columns of the Jacobian corresponding to each variable:
```
v1: (1, 1, 1e8, -1)
v2: (-1, 0, 1e10, 1-1e5)
v3: (0, -1e8, -1e6, 0)
v4: (1-1e5, 1+1e5, 1e8, -(1-1e5))
```
`v1` and `v4` are parallel in an absolute sense, but we may also want to consider `(v1, v2)` and `(v2, v4)` to be parallel pairs in a relative sense, since 1 is very small relative to the large coefficients of `1e8` and `1e10`.

See my PR into this branch, where I've implemented this relative tolerance check.

**Review by Robbybp — COMMENTED** [review]
A few small comments, then I've opened a PR into this branch where I do the following:
- Clean up variable names in some of the code I wrote for the parallel vector checker
- Implement a relative tolerance check for parallel vectors.

The PR is here: https://github.com/andrewlee94/idaes-pse/pull/19. Let me know what you think about the changes.

**Comment by andrewlee94** [comment]
@Robbybp @bknueven Do you think this PR is ready to merge now, or should we hold off? I will be out of the country from the end of the week until basically the beginning of February.

File: idaes/core/util/model_diagnostics.py
Comment (Robbybp) [inline idaes/core/util/model_diagnostics.py:3219]: I think this name is a too vague, as there are many different ways of checking for ill-conditioning. Maybe `compute_ill_conditioning_certificate` or `compute_worst_conditioned_subset`?

File: idaes/core/util/model_diagnostics.py
Comment (Robbybp) [inline idaes/core/util/model_diagnostics.py:3274]: Why is this called `inf_prob`?

File: idaes/core/util/model_diagnostics.py
Comment (Robbybp) [inline idaes/core/util/model_diagnostics.py:3323]: This does not have to happen now, but I generally like splitting these types of functions into two functions: One that simply builds the model, and one that just builds the model (by calling the first function), then solves the model. That way an advanced user can build the model and inspect it, apply a transformation, or use a custom solver. It also makes it easier to see what the default solver and options are.

I'm open to arguments that this is a bad idea, and again, this does not need to happen now.

**Review by Robbybp — COMMENTED** [review]
A few minor comments. These do not necessarily need to happen now, as this is beta functionality, but I would like them to be considered before this is made more permanent. In addition, I think a more permanent implementation should be more configurable to incorporate a DegeneracyHunter-like option.

File: idaes/core/util/model_diagnostics.py
Comment (Robbybp) [inline idaes/core/util/model_diagnostics.py:3317]: It would be nice to default to HiGHS or SCIP, although I understand if that can't happen for the time being.

File: idaes/core/util/model_diagnostics.py
Comment (bknueven) [reply idaes/core/util/model_diagnostics.py:3219]: `compute_ill_conditioning_certificate` is better.

`compute_worst_conditioned_subset` would be a bit inaccurate since it might involve solving a MIP version of this problem.

File: idaes/core/util/model_diagnostics.py
Comment (bknueven) [reply idaes/core/util/model_diagnostics.py:3323]: We actually solve two problems at the moment -- we minimize the norm of the RHS, then we fix the norm of the RHS and minimize the 1-norm of the vector. That makes a cleaner workflow more difficult.

I also imagine eventually trying to re-solve with tighter tolerances if the answer is nonsensical, e.g., the norm of the RHS is negative. These things can be done automatically much more straightforwardly if we pick a single solver.

File: idaes/core/util/model_diagnostics.py
Comment (bknueven) [reply idaes/core/util/model_diagnostics.py:3317]: The best solver to pick would probably be SoPlex as we can set arbitrary precision and utilize its iterative refinement capability. But I went with `cbc` because it ships with `idaes get-extensions`.

File: idaes/core/util/model_diagnostics.py
Comment (bknueven) [reply idaes/core/util/model_diagnostics.py:3274]: probably should just go with `m`?

Comment (Robbybp) [reply idaes/core/util/model_diagnostics.py:3274]: I like `m`. I'm fine with another name too, I just don't know what "inf" is.

File: idaes/core/util/model_diagnostics.py
Comment (Robbybp) [reply idaes/core/util/model_diagnostics.py:3219]: Since we're solving the minimum-residual problem globally, we have the "worst conditioned" combination of rows/columns, right? Then you're saying that there is no reason to believe that the result will be a subset?

File: idaes/core/util/model_diagnostics.py
Comment (Robbybp) [reply idaes/core/util/model_diagnostics.py:3323]: I did not realize that we solve two problems. I see the difficulty in splitting this up in that case.

File: idaes/core/util/model_diagnostics.py
Comment (Robbybp) [reply idaes/core/util/model_diagnostics.py:3274]: "Infeasibility" of the linear dependence constraint?

File: idaes/core/util/model_diagnostics.py
Comment (bknueven) [reply idaes/core/util/model_diagnostics.py:3219]: I don't think so, but it will tend to be a subset because we minimize the 1-norm of the multiplier alongside requiring an optimal basic feasible solution to the LP, i.e., it's just LASSO.

Put another way, we could try to minimize the number of non-zeros in the multiplier, which would give a minimal subset demonstrating the ill conditioning.

File: idaes/core/util/model_diagnostics.py
Comment (andrewlee94) [reply idaes/core/util/model_diagnostics.py:3317]: FYI, for testing at least we do have the AMPL version of SCIP available now. For DegeneracyHunter, I made the solver a user argument with SCIP as the default.

File: idaes/core/util/model_diagnostics.py
Comment (Robbybp) [inline idaes/core/util/model_diagnostics.py:3219]: I think `check_ill_conditioning` is too vague of a name here. I suggest `compute_ill_conditioning_certificate`

**Review by Robbybp — CHANGES_REQUESTED** [review]
Other than the name of `check_ill_conditioning`, I think this is ready. 

I'd still like to see this merged with DegeneracyHunter eventually, because I think having two different optimization-based degeneracy/ill-conditioning detectors will be confusing for users. This does not have to hold up this PR.

File: idaes/core/util/model_diagnostics.py
Comment (bknueven) [reply idaes/core/util/model_diagnostics.py:3219]: I like this name better.

If we were honest it might be `try_to_compute_ill_conditioning_certificate`, as this LP itself will also be ill-conditioned. SoPlex has the capability to solve LPs to arbitrary precision, but it's not available currently as a (direct) Pyomo solver. And you cannot access the same options through SCIP, unfortunately.

Comment (andrewlee94) [reply idaes/core/util/model_diagnostics.py]: I've updated the name to `compute_ill_conditioning_certificate`

**Review by Robbybp — APPROVED** [review]
Looks good

**Comment by andrewlee94** [comment]
@bknueven Would you have time to review this? I think it is just waiting on a second reviewer.

**Review by bknueven — APPROVED** [review]

**Pull request merged**
